x-pack/legacy/plugins/siem/server/lib/detection_engine/signals/utils.ts

```diff
@@ -91,3 +92,45 @@ export const getGapBetweenRuns = ({
 };
 
 export const makeFloatString = (num: number): string => Number(num).toFixed(2);
+
+/**
+ * Given a BulkResponse this will return an aggregation based on the errors if any exist
+ * from the BulkResponse. Errors are aggregated on the reason as the unique key.
+ *
+ * Example would be:
+ * {
+ *   'Parse Error': {
+ *      count: 100,
+ *      statusCode: 400,
+ *   },
+ *   'Internal server error': {
+ *       count: 3,
+ *       statusCode: 500,
+ *   }
+ * }
+ * If this does not return any errors then you will get an empty object like so: {}
+ * @param response The bulk response to aggregate based on the error message
+ * @param ignoreStatusCodes Optional array of status codes to ignore when creating aggregate error messages
+ * @returns The aggregated example as shown above.
+ */
+export const errorAggregator = (
+  response: BulkResponse,
+  ignoreStatusCodes: number[] = []
+): BulkResponseErrorAggregation => {
+  return response.items.reduce<BulkResponseErrorAggregation>((accum, item) => {
+    if (item.create.error != null && !ignoreStatusCodes.includes(item.create.status)) {
+      if (accum[item.create.error.reason] == null) {
+        accum[item.create.error.reason] = {
+          count: 1,
+          statusCode: item.create.status,
+        };
+      } else {
+        accum[item.create.error.reason] = {
```

Comment:
Thinking through permutations here: what if two errors had the same `reason` but different statuses? Should we store these by composite of status/reason instead of just reason?

I'm somewhat wary of prototype pollution here as well, since these keys are coming back over the wire. Maybe 

```ts
type BulkResponseErrorAggregation = Array<{ statusCode: number, count: number, message: string }>
```
 ?

Reply:
It would be very weird if they were the same reason but different status. I think that is an impossibility from Elastic Search?

Reply:
I don't think this is going to have issues with prototype pollution as I don't do deep merges but to keep the chance non-existent the simplest trick that also saves memory is to use:

```ts
Obect.create(null)
```

In the initialization of the object. Ref: https://codeburst.io/what-is-prototype-pollution-49482fc4b638

That makes an object without a prototype and also it saves a bit in memory.